[자동차 경주 -3,4단계] 안석환 미션 제출합니다

seokhwan-an opened the pull request:
### 이번 미션을 진행하면서 집중한 점
1. model, view를 분리하고 model이 view에 의존하지 않도록 했습니다.
2. 사용자 입력에 대해서 추가적인 예외사항을 추가했습니다.(자동차 이름 중복)

코드 확인해주시고 잘못되었거나 실수한 점 있으면 편하게 리뷰 부탁드립니다!🙇‍♂️

seokhwan-an committed:
docs: 3단계 요구사항 정리

seokhwan-an committed:
style: view 부분 추가하기 전에 package 분리

seokhwan-an committed:
feat: 자동차 이름을 입력받는 기능 추가

seokhwan-an committed:
feat: 자동차 이름을 입력받는 기능 추가

seokhwan-an committed:
feat: 자동차 이름을 검증하는 기능 추가

세부사항
- 자동차 이름은 null이 될 수 없다.
- 자동차 이름은 빈칸이 될 수 없다.
- 자동차 이름은 5자 이하이다.

seokhwan-an committed:
feat: 경주 시도횟수를 검증한다.

세부사항
- 경주 시도횟수가 0보다 큰 수여야 한다.

seokhwan-an committed:
feat: 경주 진행사항을 출력한다.

seokhwan-an committed:
feat: 우승자를 출력한다.

seokhwan-an committed:
feat: 자동차 경주 실행 추가

seokhwan-an committed:
feat: 중복된 자동차 이름 검증 추가

seokhwan-an committed:
style: 불필요한 import 제거 및 코드 포맷 수정

seokhwan-an committed:
refactor: 잘못된 메소드 사용 수정

somefood commented:
안녕하세요 석환님
3, 4단계 진행하시느라 수고 많으셨습니다.
확인 한 번 해보고 궁금한 점 여쭙도록 하겠습니다~! :)

somefood commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+    private static final String PROGRESS_FORMAT = "-";
+    private static final String WINNER_DELIMITER = ", ";
+
+    public static void printRacing(final List<Car> cars) {
+        for (Car car : cars) {
+            System.out.printf("\n%s : %s", car.getName(), getProgressVar(car.getPosition()));
+        }
+        System.out.println();
+    }
+
+    private static String getProgressVar(int position) {
+        return PROGRESS_FORMAT.repeat(position);
+    }
+
+    public static void printResult(final List<Car> cars) {
+        System.out.printf("\n%s가 최종 우승했습니다.", getWinnerNames(cars));
+    }
```
자동차의 현재 위치를 출력하는 역할을 해당 클래스에 맡기신 이유가 궁금합니다!

seokhwan-an replied:
자동차는 domain으로 자동차의 position 상태를 변경하는 것에만 집중했고 OutputView에서는 출력에 대한 책임을 처리하는데 집중했습니다. 

이렇게 한 이유는 만약 자동차에서 출력 부분까지 책임을 지고 있다면 출력형태가 변경되거나 출력 여부가 변경되게 될 때 (view의 방식이 변경될 때) view의 코드가 수정되는 것이 아닌 domain의 코드가 수정되는 상황이 발생하는 것을 막기 위함이었습니다.  

그래서 자동차의 현재 위치를 출력하는 것은 자동차 내부에서 하는 것이 아닌 OutputView에서 처리하도록 했습니다.

somefood replied:
아하 
OutputView에 출력에 대한 책임을 처리함을 이해했습니다.
출력이 -가 아니라 다른 형태로 나타날 수 있으니, 자동차에겐 위치 출력을 안 하도록 하는 거군요

궁금한 것이 자동차에게 위치 출력에 대한 책임까지 부여하면 너무 많은 책임을 지게 하는 것일까요?
형태가 달라지는 출력이라도 자동차가 position이란 정보를 캡슐화해서 갖고있으면 자동차에게 출력에 대한 책임을 줘도 괜찮다고 생각하는데, 석환님의 생각은 어떠신지 궁금합니다 ㅎㅎ

또한 자동차 객체로만 단독으로 상태를 보고 싶은 경우도 생기지 않을까란 생각도 들어서 이렇게 여쭙게 됩니다~!

(근데 또 한편으론 저희가 getPosition이란 메서드를 이미 public 하게 오픈했으니 이를 활용해서 출력해도 또 문제가 없을거 같군요... 객체지향은 넘모 어렵습니다 ㅠㅠ)

seokhwan-an replied:
제가 잘 이해를 하지 못한 것 같은데 그러면 석주님은 position을 공개하는 것을 캡슐화로 막고 출력 책임을 자동차로 한다는 것을 의미하는 것인가요?

이렇게 되면 model과 view의 역할이 서로 혼용되는 것 같은데 그 부분에 대해서는 어떻게 생각하시나요?

seokhwan-an replied:
제가 객체 지향 공부를 하면서 느낀 캡슐화는 객체의 상태를 외부에 노출시키지 않는 것이 아니라 외부에서 상태에 직접 접근해서 변경하는 작업을 하지 못하게 하는 것이라고 느꼈습니다.(필요한 부분에 대해서는 외부에 노출 시켜도 괜찮다고 생각합니다. 이번 상황은 view에 Position 정보를 제공해주는 것이 있을 것 같습니다.) 석주님은 position을 외부에 노출하는 것이 문제라고 생각하시는 것일까요?

seokhwan-an replied:
"또한 자동차 객체로만 단독으로 상태를 보고 싶은 경우도 생기지 않을까란 생각도 들어서 이렇게 여쭙게 됩니다~!" 이 부분의 의미가 명확하게 파악이 안되는데 혹시 특정 한 자동차에 대한 정보만 필요한 경우를 의미하는 것이라면 출력을 담당하는 객체에 한 대의 자동차 정보를 input으로 받아서 출력하는 기능을 추가해서 단독 정보를 볼 수 있을 것 같습니다.(이를 domain에서 직접 출력할 필요가 없다고 생각합니다) 혹시 잘못 이해한 것이라면 다시 의견 달아주세요!

seokhwan-an replied:
미션 요구사항을 보면 크게 2가지로 분리할 수 있을 것 같습니다. 비즈니스 로직에 대한 요구사항 (ex  0~9 중 랜덤한 값으로 4이상일 경우에만 전진한다.)과 화면에 대한 요구사항(ex 경주 결과를 나타낸다, 경주 상황을 나타낸다)으로 분리될 수 있고 비즈니스 로직에 대한 책임은 domain이 화면에 대한 책임은 view가 가지고 있습니다.

 위의 궁금해 하신 것 중 자동차 위치 출력에 대한 책임을 자동차에 부여하는 것은 너무 많은 책임을 부여하는 것인지 물어보셨는데 유지보수 관점에서 현재는 프로그램의 규모가 작기 때문에 직접 처리하더라도 유지보수가 어렵지 않을 것 같은데 프로그램이 점점 커지게 되어서 domain이 view에 대한 처리까지 책임을 진다면 이 둘을 분리해서 발전해 나가는 것이 어려울 것 같습니다!

somefood replied:
석환님 친절하고 상세한 답변 감사드립니다.

이미 석환님께서 답글을 잘 달아주셨지만 질문들에 답변을 드리자면,

Q) "position을 공개하는 것을 캡슐화로 막고 출력 책임을 자동차로 한다는 것을 의미하는 것인가요?"
A) 막자는 것은 아니었고, 상태에 대한 출력을 자동차에서 처리해도 되지 않을까라는 의문이 들었던 거였습니다!

Q) "position을 외부에 노출하는 것이 문제라고 생각하시는 것일까요?"
A) 저 또한 이미 많은 개발을 할 때 자바 빈 프로퍼티즈를(Getter 정도만) 적극 활용하고 있기에 석환님 말씀대로 필요한 부분은 노출해도 된다고 생각하고 있습니다. 

Q) "model과 view의 역할이 서로 혼용되는 것 같은데 그 부분에 대해서는 어떻게 생각하시나요?"
A) 요 부분은 제가 MVC 패턴의 이해가 덜 된듯 합니다. ㅠㅠ

어찌되었건, 제가 위 질문에서 궁금했던 요지는 (position의 공개 여부 상관없이) 자동차라는 객체에서 상태 출력해도 괜찮지 않을까 였는데, 석환님의 말씀대로 비즈니스 로직과 출력 부분을 두 개의 관심사로 바라보아, 분리하는게 이번 MVC 패턴의 내용인거 같군요 ㅎㅎ

다음 미션에서는 저도 관심사를 좀 더 분리해보도록 해보겠습니다. 
감사합니다~!

somefood commented on src/main/java/model/Name.java:
자동차의 이름을 관리하는 클래스를 통해 유효성 검사를 하는 것이 좋았습니다! ㅎㅎ

seokhwan-an replied:
감사합니다!

somefood commented on src/main/java/model/Cars.java:
```diff
@@ -0,0 +1,49 @@
+        return new Cars(cars);
+    }
+
+    private static void validateDuplicatedName(final String[] carNames) {
+        final Set<String> notDuplicatedNames = new HashSet<>(Arrays.asList(carNames));
+        if (notDuplicatedNames.size() != carNames.length) {
+            throw new IllegalArgumentException("중복된 이름을 가진 차들이 존재합니다.");
+        }
+    }
```
요구사항에선 이름 중복에 대한 처리는 없는거 같은데,
중복 검사를 하는 이유가 있을까요~?

seokhwan-an replied:
이 부분은 제가 추가적으로 생각해서 부여한 제약사항입니다. 참여하는 자동차가 같은 이름을 가지게 되면 누가 우승을 했는지 파악하는게 어렵다고 판단했습니다.

somefood replied:
그렇군요.
반대로 이름이 중복 가능한 경우라면 석환님은 어떤식으로 처리하실지 개인적으로 궁금합니다 ㅎㅎ

seokhwan-an replied:
중복된 이름을 가진다면 내부적으로 uuid를 부여해서 이 둘을 식별할 것 같습니다!

somefood replied:
UUID 좋은 아이디어인거 같습니다! ㅎㅎ

somefood commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,22 @@
+        OutputView.printResult(racingGame.getWinners());
+    }
+
+    private static void race(final RacingGame racingGame) {
+        for (int i = 0; i < racingGame.getRacingCount(); i++) {
+            racingGame.moveCars();
+            OutputView.printRacing(racingGame.getCars());
+        }
+    }
```
racingCount 만큼 자동차를 이동시키는 역할을 RacingGame 클래스 안에서 해도 될거 같은데, main에서 움직이게 한 이유가 궁금합니다!

seokhwan-an replied:
이 부분도 앞선 응답과 같이 domain에서 view를 의존하는 상황을 막기 위해서 main에서 처리했습니다.

somefood replied:
넵 의도하신 내용은 알 거 같습니다!

https://github.com/next-step/java-racingcar-simple-playground/pull/41/files#r1618009236

상기 링크에서 석환님이 말씀주신대로 코드를 이해했습니다.

한편으로는 저는 racingGame 객체에서 racingCount를 내부적으로 캡슐화해서 가지고 있는데,
이것을 애플리케이션단에서 꺼내서 자동차의 움직임을 반복하는게 맞을까라는 의문이 생겼거든요.

캡슐화에 대해선 어떻게 생각하시는지 의견 여쭤보고 싶습니다!

seokhwan-an replied:
저도 이 부분은 코드가 객체 지향적인 코드라고 생각하지 않습니다! 하지만 현재 view와 model이 함께 이용되는 코드여서 현재 controller 역할을 담당하는 Application 클래스에서 처리를 했습니다.

캡슐화에 대한 제 의견은 위에 적어 두었습니다!

somefood commented on src/test/java/model/RacingGameTest.java:
```diff
@@ -11,29 +12,13 @@
     void get_winners(Cars cars, List<Car> expect) {
         // given
         final int count = 5;
-        final RacingGame racingGame = new RacingGame(cars, count);
+        final RacingGame racingGame = new RacingGame(cars, new RacingCount(count));
 
         // when
         racingGame.moveCars();
```
racingGame.moveCars()를 보았을 땐 한 번만 자동차들을 움직이게 하여서
racingCount의 활용이 안 되고 있는 거 같습니다.
카운트가 n일 때 자동차의 position이 잘 작동하고 있는지 확인이 있으면 좋을거 같습니다! ㅎㅎ

seokhwan-an replied:
넵 이 부분은 추가하겠습니다!

seokhwan-an committed:
test: n번 동안 경주 테스트 추가

세부사항
- 테스트 용이성을 위해 RacingGameController 추가
- 여러 횟수에 따른 자동차 움직임 제어를 위해 ListMoveStrategy 추가

somefood reviewed (approved):
확인했습니다.
구현하느라 수고 많으셨습니다~! :)
